### PR TITLE
frontend: add service registry

### PR DESCRIPTION
**Pull request by stefanhengl**
This adds a tiny service registry to frontend. 

## Why?
- We want to add a "re-index" button to the index status page.
- Frontend doesn't know the addresses of the Zoekt indexservers.
- Other projects that could potentially benefit: 1-click export, health-checker

## Goals
- Create a simple but resilliant service registry
- Support all environments Sourcegraph can run on (server, docker-compose, K8S, other containerized environments) 
- Don't require open ports other than those which are already open
- Don't add another service
- Don't create too much load on frontend

## Non-goals
- Make this a general purose service registry that is useful beyond Sourcegraph

## Not part of this PR
- a client for the service registry. Once/if this is merged, we can create a separate repo (sourcegraph/registry-client) that both Sourcegraph and Zoekt can import. 

## Design
image

## Test plan
- local testing
```bash
# create record
 curl -v -X POST http://localhost:3090/.internal/services/foobar -d '{"port": 8119, "health_check_path": "/healthz", "hostname": "host1"}'

# renew
curl -v -X PUT http://localhost:3090/.internal/services/foobar/127.0.0.1:8119

# delete
curl -v -X DELETE http://localhost:3090/.internal/services/foobar/127.0.0.1:8119
```
- Checked that manually created entries expire if they are not renewed
- Wrote a basic client for Zoekt (not part of this PR) and verified that db records show up as expected.

### CODE REVIEW

**Comment by keegancsmith** [comment]
Sorry haven't looked at the code, but at the excellent description. The design looks good to me. Our scale is small, so a simple registry in a postgres table sounds like is the correct trade off vs the more complicated solutions out there.

I wonder if we could remove our existing usages of the k8s API to do service discovery? Right now I'm not exactly sure we could, because we prefer more "static" service discovery (ie just reading the replica count on the stateful set).

Thinking specifically about the health check tool, if the DB is down then things break. But that seems fine, the health check tool saying the DB is down/not writeable is good information of the most important thing to fix :)

I won't be able to take a closer look until I get back from PTO, so maybe ask for a review in #dev-backend?

**Comment by stefanhengl** [comment]
@ggilmore could you take a look? We can also go over it in one of our sync meetings if you prefer.

File: internal/database/schema.md [hunk 1]
```diff
@@ -2474,6 +2474,26 @@ Contains security-relevant events with a long time horizon for storage.
 
 **version**: The version of Sourcegraph which generated the event.
 
+# Table "public.service_registry"
+```
+      Column       |           Type           | Collation | Nullable | Default 
+-------------------+--------------------------+-----------+----------+---------
+ ip                | text                     |           | not null | 
+ port              | integer                  |           | not null |
```
Comment (ggilmore) [inline internal/database/schema.md:2482]: As discussed during the sync, it's possible for the following to happen:

- `zoekt-indexserver-0` with ip/port `192.168.1.1:6666` registers itself with the sourcegraph-frontend instance 
- `zoekt-indexserver-0` is killed without being able to de-register from frontend
- - `zoekt-indexserver-0` is restarted with a different ip/port:   `192.168.2.2:6666` and registers registers itself with sourcegraph-frontend

Now the registry will have two entries for things that both say "service: zoekt-indexserver" without any way to disambiguate  them (these entries will look similar to entries provided by zoekt-indexserver-1,2,3,...). 


---

What's missing is a way to determine if we have two entries for the same _replica_. If we knew the _hostname_ associated with each entry, the client can implement the logic to only use the latest value for a given hostname. I don't know if we need a hostname column or just a generic "metadata" one, but in either way I think we need more state to handle this case.

Comment (stefanhengl) [reply internal/database/schema.md:2564]: I added the hostname as a required field.

**Comment by keegancsmith** [comment]
@jac we mentioned this PR while on the onsite. Do you wanna take a look? I'm happy to review and land this so you can get started on the health checker tool.

**Comment by jac** [comment]
Ah, nice! This should also be useful for https://github.com/sourcegraph/sourcegraph/issues/20249 which has come up a few times recently

**Comment by stefanhengl** [comment]
closing this one, because it didn't gain enough traction. @jac feel free to reopen or just copy/paste if you find something useful.

**Comment by stefanhengl** [comment]
Attempt #2: This has become relevant again in the context of adding a "re-index" button to the index status page.

**Review by keegancsmith — CHANGES_REQUESTED** [review]
I'm unsure if this is ready for review again, so just requesting changes to get out of my queue.

**Pull request closed without merging**